dlt/destinations/impl/lancedb/lancedb_client.py

```diff
@@ -0,0 +1,698 @@
+)
+from dlt.destinations.impl.lancedb.lancedb_adapter import VECTORIZE_HINT
+from dlt.destinations.impl.lancedb.schema import (
+    arrow_schema_to_dict,
+    make_arrow_field_schema,
+    make_arrow_table_schema,
+    TArrowSchema,
+    NULL_SCHEMA,
+    TArrowField,
+)
+from dlt.destinations.impl.lancedb.utils import (
+    list_unique_identifiers,
+    generate_uuid,
+    set_non_standard_providers_environment_variables,
+)
+from dlt.destinations.job_impl import EmptyLoadJob
+from dlt.destinations.type_mapping import TypeMapper
+
+
+TIMESTAMP_PRECISION_TO_UNIT: Dict[int, str] = {0: "s", 3: "ms", 6: "us", 9: "ns"}
+UNIT_TO_TIMESTAMP_PRECISION: Dict[str, int] = {v: k for k, v in TIMESTAMP_PRECISION_TO_UNIT.items()}
+
+
+class LanceDBTypeMapper(TypeMapper):
+    sct_to_unbound_dbt = {
+        "text": pa.string(),
+        "double": pa.float64(),
+        "bool": pa.bool_(),
+        "bigint": pa.int64(),
+        "binary": pa.binary(),
+        "date": pa.date32(),
+        "complex": pa.string(),
+    }
+
+    sct_to_dbt = {}
+
+    dbt_to_sct = {
+        pa.string(): "text",
+        pa.float64(): "double",
+        pa.bool_(): "bool",
+        pa.int64(): "bigint",
+        pa.binary(): "binary",
+        pa.date32(): "date",
+    }
+
+    def to_db_decimal_type(
+        self, precision: Optional[int], scale: Optional[int]
+    ) -> pa.Decimal128Type:
+        precision, scale = self.decimal_precision(precision, scale)
+        return pa.decimal128(precision, scale)
+
+    def to_db_datetime_type(
+        self, precision: Optional[int], table_format: TTableFormat = None
+    ) -> pa.TimestampType:
+        unit: str = TIMESTAMP_PRECISION_TO_UNIT[self.capabilities.timestamp_precision]
+        return pa.timestamp(unit, "UTC")
+
+    def to_db_time_type(
+        self, precision: Optional[int], table_format: TTableFormat = None
+    ) -> pa.Time64Type:
+        unit: str = TIMESTAMP_PRECISION_TO_UNIT[self.capabilities.timestamp_precision]
+        return pa.time64(unit)
+
+    def from_db_type(
+        self,
+        db_type: pa.DataType,
+        precision: Optional[int] = None,
+        scale: Optional[int] = None,
+    ) -> TColumnType:
+        if isinstance(db_type, pa.TimestampType):
+            return dict(
+                data_type="timestamp",
+                precision=UNIT_TO_TIMESTAMP_PRECISION[db_type.unit],
+                scale=scale,
+            )
+        if isinstance(db_type, pa.Time64Type):
+            return dict(
+                data_type="time",
+                precision=UNIT_TO_TIMESTAMP_PRECISION[db_type.unit],
+                scale=scale,
+            )
+        if isinstance(db_type, pa.Decimal128Type):
+            precision, scale = db_type.precision, db_type.scale
+            if (precision, scale) == self.capabilities.wei_precision:
+                return cast(TColumnType, dict(data_type="wei"))
+            return dict(data_type="decimal", precision=precision, scale=scale)
+        return cast(
```

Comment:
why not call super here? this code is implemented there isn't it?

Reply:
I overrode `from_dbt_type` as associated methods because we have to return a `pa.DataType` instance for `LanceDBTypeMapper` to work correctly.

In addition to this, the timestamp precision in pyarrow has to be accounted for, the constructors don't just take a scalar but a string. Please see the unit parameter at https://arrow.apache.org/docs/python/generated/pyarrow.timestamp.html

Reply:
```python
...
        return cast(
            TColumnType,
            without_none(
                dict(
                    data_type=self.dbt_to_sct.get(db_type, "text"),
                    precision=precision,
                    scale=scale,
                )
            ),
        )
```

I did replace the above with `return super().from_db_type(db_type, precision, scale)`  though to reduce redundancy. Thanks for the spot!

Reply:
yes, that's whay I meant I think, thanks